### PR TITLE
STELLAR-1044 Replace OS pipe in AR2300 driver with blocking queue.

### PR DESCRIPTION
**Pull request by reiinakano**
Replaces the OS pipe with a special queue that blocks during `pop` until data is available.

### CODE REVIEW

File: gr-starcoder/lib/starcoder_queue.cc [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+    q_.push(arr[i]);
+  }
+  if (was_empty) {
+    condition_var_.notify_one();
+  }
+  return size;
+}
+
+size_t starcoder_queue::pop(char *arr, size_t size) {
+  boost::mutex::scoped_lock lock(mutex_);
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: I think this prevents any more `push` while `pop` is waiting on the condition variable. Since we only have a single consumer, I think it's safe to have the lock after the wait.

File: gr-starcoder/lib/starcoder_queue.cc [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  int i = 0;
+  while (i < size && !q_.empty()) {
+    arr[i] = q_.front();
+    q_.pop();
+    i ++;
+  }
+  return i;
+}
+
+size_t cpp_callback_push(starcoder_queue* q, const char *arr, size_t size) {
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: This isn't really used as a callback, I think just `pushToQueue` or similar semantic naming is fine.

File: gr-starcoder/lib/starcoder_queue.cc [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this software; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "starcoder_queue.h"
+
+starcoder_queue::starcoder_queue() :
+  q_(internal_ring_buffer(boost::circular_buffer<char>(1048576)))
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: Pass the size as a constructor parameter.

File: gr-starcoder/lib/starcoder_queue.cc [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+  boost::mutex::scoped_lock lock(mutex_);
+  while (q_.empty()) {
+    condition_var_.wait(lock);
+  }
+
+  int i = 0;
+  while (i < size && !q_.empty()) {
+    arr[i] = q_.front();
+    q_.pop();
+    i ++;
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: `i++`

File: gr-starcoder/lib/starcoder_queue.cc [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+starcoder_queue::starcoder_queue() :
+  q_(internal_ring_buffer(boost::circular_buffer<char>(1048576)))
+{ }
+
+size_t starcoder_queue::push(const char *arr, size_t size) {
+  boost::mutex::scoped_lock lock(mutex_);
+  bool const was_empty = q_.empty();
+  for (int i=0; i < size; i++) {
+    q_.push(arr[i]);
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: Is this not the right documentation? I see `push_back` and `pop_front` instead of `push`, `pop`.
https://www.boost.org/doc/libs/1_66_0/boost/circular_buffer/base.hpp

File: gr-starcoder/lib/starcoder_queue.cc [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+#include "starcoder_queue.h"
+
+starcoder_queue::starcoder_queue() :
+  q_(internal_ring_buffer(boost::circular_buffer<char>(1048576)))
+{ }
+
+size_t starcoder_queue::push(const char *arr, size_t size) {
+  boost::mutex::scoped_lock lock(mutex_);
+  bool const was_empty = q_.empty();
+  for (int i=0; i < size; i++) {
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: `i = 0`

File: gr-starcoder/lib/starcoder_queue.cc (hunk 1 above)
Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.cc]: The call to `condition_var_.wait(lock);` below will atomically unlock the mutex while waiting. I used this reference for this https://www.justsoftwaresolutions.co.uk/threading/implementing-a-thread-safe-queue-using-condition-variables.html

File: gr-starcoder/lib/starcoder_queue.cc (hunk 5 above)
Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.cc]: I'm using the `std::queue` as a container over `circular_buffer` http://en.cppreference.com/w/cpp/container/queue

File: gr-starcoder/lib/ar2300_driver.h [hunk 7]
```diff
@@ -187,14 +187,9 @@ int ar2300_start_transfer(AR2300_HANDLE *ar2300);
...
- * set the output file descriptor
- *
- * set the fd to which the output will be written to
- *
- * @param ar2300 handle to the ar2300 device
- * @param fd the file to write I/Q data to
+ * set the queue to write IQ packets to
  */
-void ar2300_set_fd(AR2300_HANDLE *ar2300, int fd);
+void ar2300_set_q(AR2300_HANDLE *ar2300, starcoder_queue *q);
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/ar2300_driver.h]: Not sure `q` is an accepted common abbreviation of `queue`. Spell out as `ar2300_set_queue`.

File: gr-starcoder/lib/ar2300_receiver.h [hunk 8]
```diff
@@ -51,6 +52,9 @@ class ar2300_receiver {
   // Pipe for reading data from AR2300
   int read_pipe[2];
 
+  // Blocking queue
+  starcoder_queue q;
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/ar2300_receiver.h]: same as above

File: gr-starcoder/lib/ar2300_driver.h [hunk 9]
```diff
@@ -115,7 +115,7 @@ typedef struct ar2300_handle {
 
   ar2300_bulk_status bulk_status; /**< status of bulk transfer */
 
-  int outfd; /**< I/Q output file descriptor */
+  starcoder_queue* q; // Blocking queue for storing IQ data
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/ar2300_driver.h]: spell out `queue` as I explained below.

File: gr-starcoder/lib/starcoder_queue.h [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+#include <boost/thread/mutex.hpp>
+  typedef std::queue<char, boost::circular_buffer<char>> internal_ring_buffer;
+  class starcoder_queue {
+  public:
+    starcoder_queue();
+    size_t push(const char*, size_t);
+    size_t pop(char*, size_t);
+  private:
+    internal_ring_buffer q_;
+    boost::fibers::condition_variable_any condition_var_;
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: use standard mutex and condition variable classes of C++11.
http://en.cppreference.com/w/cpp/thread/mutex
http://en.cppreference.com/w/cpp/thread/condition_variable

**Comment by reiinakano** [comment]
Comments addressed

File: gr-starcoder/lib/starcoder_queue.cc [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+ */
+
+#include "starcoder_queue.h"
+
+starcoder_queue::starcoder_queue(int buffer_size) :
+  q_(internal_ring_buffer(boost::circular_buffer<char>(buffer_size)))
+{ }
+
+size_t starcoder_queue::push(const char *arr, size_t size) {
+  std::unique_lock<std::mutex> lock(mutex_);
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: How about using `std::defer_lock`?

File: gr-starcoder/lib/starcoder_queue.cc [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+  return size;
+}
+
+size_t starcoder_queue::pop(char *arr, size_t size) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (q_.empty()) {
+    condition_var_.wait(lock);
+  }
+
+  int i = 0;
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.cc]: Let's name this `popped`

File: gr-starcoder/lib/starcoder_queue.cc [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "starcoder_queue.h"
+
+starcoder_queue::starcoder_queue(int buffer_size) :
+  q_(internal_ring_buffer(boost::circular_buffer<char>(buffer_size)))
+{ }
+
+size_t starcoder_queue::push(const char *arr, size_t size) {
```
Comment (gavinreaney) [inline gr-starcoder/lib/starcoder_queue.cc]: If the circular buffer is full, I think this will overwrite the data at the start of the buffer. Is that what we want? I guess the previous code would block. Neither approach is very ideal but I think we'd want to know if this condition happens.

File: gr-starcoder/lib/starcoder_queue.cc [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+ * You should have received a copy of the GNU General Public License
+ * along with this software; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "starcoder_queue.h"
+
+starcoder_queue::starcoder_queue(int buffer_size) :
+  q_(internal_ring_buffer(boost::circular_buffer<char>(buffer_size)))
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.cc]: (It was long before when I was actively using C++, so please disregard if my comment is pointless.)

Can you write 
`q_(boost::circular_buffer<char>(buffer_size))`
?

File: gr-starcoder/lib/starcoder_queue.h [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+#include <mutex>
+#include <condition_variable>
+  typedef std::queue<char, boost::circular_buffer<char>> internal_ring_buffer;
+  class starcoder_queue {
+  public:
+    starcoder_queue(int buffer_size);
+    size_t push(const char*, size_t);
+    size_t pop(char*, size_t);
+  private:
+    internal_ring_buffer q_;
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: `queue_`. Use `q` only for short-lived local variables.

File: gr-starcoder/lib/starcoder_queue.h [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+
+/* This header can be read by both C and C++ compilers */
+#ifndef STARCODER_QUEUE_H
+#define STARCODER_QUEUE_H
+#ifdef __cplusplus
+#include <queue>
+#include <boost/circular_buffer.hpp>
+#include <mutex>
+#include <condition_variable>
+  typedef std::queue<char, boost::circular_buffer<char>> internal_ring_buffer;
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: no indent this code block.

File: gr-starcoder/lib/starcoder_queue.h [hunk 17]
```diff
@@ -0,0 +1,52 @@
...
+  };
+#else
+  typedef
+    struct starcoder_queue
+      starcoder_queue;
+#endif
+#ifdef __cplusplus
+extern "C" {
+#endif
+  extern size_t push_to_queue(starcoder_queue*, const char*, size_t);
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: no indent.  Preprocessor conditionals don't typically create indentation.

Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: Move those two functions above the class definition. The public interface exposed by this file is the functions and the class is implementation detail.  A rule of thumb is public interface comes first and then implementation details follow in a file.

Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: Do not omit argument names in forward declarations. They may play a great role of communicating purpose/meaning of those arguments. Without them, a user of this code has no idea what they should pass into those functions.

File: gr-starcoder/lib/ar2300_receiver.cc [hunk 18]
```diff
@@ -136,49 +138,17 @@ void ar2300_receiver::stop() {
...
-    return 0;
-  }
-
-  int ret = ::read(read_pipe[PIPE_READ], buf, size);
-  if (ret < 0) {
-    stop();
-    fprintf(stderr, "ar2300_receiver::read: failed to read data from the device. ret=%d\n", ret);
-    throw std::runtime_error("ar2300_receiver::read");
-  }
+  int ret = queue_.pop(buf, size);
```
[the remark this answers is not in the document]
Comment (reiinakano) [reply gr-starcoder/lib/ar2300_receiver.cc]: Good catch. I initially thought it would be fine, since https://wiki.gnuradio.org/index.php/BlocksCodingGuide says that condition variable `wait`s are fine to block indefinitely. When I tested it out though, I couldn't stop the flowgraph by simply calling `stop`. So no harm keeping the timeout.

File: gr-starcoder/lib/starcoder_queue.h (hunk 17 above)
Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.h]: If I move it to the top, the compiler seems to error out since I haven't declared it yet.

File: gr-starcoder/lib/starcoder_queue.cc [hunk 19]
```diff
@@ -0,0 +1,59 @@
...
+{ }
+
+size_t starcoder_queue::push(const char *arr, size_t size) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  bool const was_empty = q_.empty();
+  for (int i = 0; i < size; i++) {
```
[the remark this answers is not in the document]
Comment (kenichi-fukushima) [reply gr-starcoder/lib/starcoder_queue.cc]: +1

File: gr-starcoder/lib/starcoder_queue.cc [hunk 20]
```diff
@@ -0,0 +1,60 @@
...
+  if (was_empty) {
+    condition_var_.notify_one();
+  }
+  return size;
+}
+
+size_t starcoder_queue::pop(char *arr, size_t size, int timeout_ms) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  while (queue_.empty()) {
+    condition_var_.wait_for(lock, std::chrono::milliseconds(timeout_ms));
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.cc]: The documentation says the thread acquires lock when this function returns for whatever reasons including timeout, so it is safe to proceed. However, it is surprising to me that the code proceeds when the pre-condition is not met. Please either leave a comment that you intentionally proceed regardless of the return value, or check the return value and do early return when timeout.

File: gr-starcoder/lib/starcoder_queue.h [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+#include <mutex>
+#include <condition_variable>
+#include <boost/lockfree/spsc_queue.hpp>
+class starcoder_queue {
+  public:
+    starcoder_queue(int buffer_size);
+    size_t push(const char*, size_t);
+    size_t pop(char*, size_t, int);
+  private:
+    boost::lockfree::spsc_queue<char> queue_;
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: Nice. `spsc_queue` is thread-safe so you can get rid of mutex and condvar.
https://www.boost.org/doc/libs/1_53_0/doc/html/boost/lockfree/spsc_queue.html#idp67134384-bb

Everything in boost::lockfree seems to be thread-safe.
https://www.boost.org/doc/libs/1_66_0/doc/html/lockfree.html

File: gr-starcoder/lib/starcoder_queue.h [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+ * along with this software; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+/* This header can be read by both C and C++ compilers */
+#ifndef STARCODER_QUEUE_H
+#define STARCODER_QUEUE_H
+#ifdef __cplusplus
+#include <queue>
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/starcoder_queue.h]: Remove this include.

File: gr-starcoder/lib/starcoder_queue.cc (hunk 19 above)
[the remark this answers is not in the document]
Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.cc]: I changed the implementation from `std::queue` to `spsc_queue`. This is already faster than the previous implementation (no precise benchmarks, but the CPU usage reported by system monitor went from 2-4% to 0-1%). The code is also much cleaner. I considered just `circular_buffer` instead of `spsc_queue` but I found this discussion: https://stackoverflow.com/questions/35675261/why-is-boostcircular-buffer-so-slow-in-my-benchmark

File: gr-starcoder/lib/starcoder_queue.cc (hunk 13 above)
Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.cc]: `spsc_queue` now won't write more than the available space left in the buffer

File: gr-starcoder/lib/starcoder_queue.h (hunk 21 above)
Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.h]: Actually, the very first implementation of this was written with `spsc_queue`. I need this to be a blocking queue actually, since GNURadio will just call `pop` repeatedly and cause 100% CPU usage. So now the solution is combination of `spsc_queue` and the mutexes.

File: gr-starcoder/lib/starcoder_queue.cc (hunk 20 above)
Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.cc]: Oops. Good catch. This was a bug

File: gr-starcoder/lib/starcoder_queue.h (hunk 21 above)
Comment (kenichi-fukushima) [reply gr-starcoder/lib/starcoder_queue.h]: Ah, got it. Then, please leave a comment. Basically, any decisions made by the code author that are not obvious to other people is worth commenting.

Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.h]: Never mind, I can get rid of the mutex but keep the `cond_variable` I think. Let me do some more reading..

File: gr-starcoder/lib/starcoder_queue.h [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+class starcoder_queue {
+  public:
+    starcoder_queue(int buffer_size);
+    size_t push(const char*, size_t);
+    size_t pop(char*, size_t, int);
+  private:
+    // Although we use the threadsafe spsc_queue here, we still need to keep the
+    // mutex for the sole purpose of waiting for the queue to be non-empty.
+    // This is since GNURadio will repeatedly call the non-blocking pop and
+    // consume 100% CPU.
```
Comment (anuraaga) [inline gr-starcoder/lib/starcoder_queue.h]: I think more accurately this is because AR2300 samples at 1Mhz while our CPUs are at Ghzs, so there is a lot of blocking. Perhaps we will have faster sampling in the future where spin lock is more efficient than a condition var - do you think we should rename this so it's not so general as to preclude such a case? If not would at least document the sampling rate relationship

Comment (reiinakano) [reply gr-starcoder/lib/starcoder_queue.h]: Yeah, the queue is almost always empty because of this. Anyway, by the time our sampling rates are comparable to our CPU rates, I doubt this code would still be relevant. 

So yeah maybe `starcoder_queue` is too general. What do you think about `blocking_queue`? Weird that Boost or the standard lib doesn't already have this.

Comment (anuraaga) [reply gr-starcoder/lib/starcoder_queue.h]: Java has blocking queue but lock-free is a library called disruptor - I guess both have different philosphies ;)

`blocking_queue` SGTM

File: gr-starcoder/lib/ar2300_source_impl.cc [hunk 24]
```diff
@@ -42,9 +42,9 @@ namespace gr {
       : gr::sync_block("ar2300_source",
               gr::io_signature::make(0, 0, 0),
               gr::io_signature::make(1, 1, sizeof(gr_complex))),
-        receiver(new ar2300_receiver())
+        receiver(new ar2300_receiver(10485760)), // 10MB buffer
+        timeout_ms(10000)
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/ar2300_source_impl.cc]: The previous value was 1000. Is this change intentional?

File: gr-starcoder/lib/blocking_queue.h [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+};
+#else
+typedef
+  struct blocking_queue
+    blocking_queue;
+#endif
+#ifdef __cplusplus
+extern "C" {
+#endif
+extern size_t push_to_queue(blocking_queue* q, const char* arr, size_t size);
```
Comment (kenichi-fukushima) [inline gr-starcoder/lib/blocking_queue.h]: (nit/optional) A naming convention in C API design is to name a function after the main data structure that it operates on.  Because C doesn't have such concepts as namespace or package, this is the only way to avoid name collision and improve code readability. In this case, you could name this function `blocking_queue_push` and the below `blocking_queue_pop`.

File: gr-starcoder/lib/blocking_queue.cc [hunk 26]
```diff
@@ -0,0 +1,56 @@
...
+
+blocking_queue::blocking_queue(int buffer_size) :
+  queue_(buffer_size)
+{ }
+
+size_t blocking_queue::push(const char *arr, size_t size) {
+  std::unique_lock<std::mutex> lock(mutex_);
+  bool const was_empty = queue_.empty();
+  int pushed = queue_.push(arr, size);
+  if (was_empty) {
```
Comment (anuraaga) [inline gr-starcoder/lib/blocking_queue.cc]: `pushed > 0 && was_empty`

File: gr-starcoder/lib/blocking_queue.cc [hunk 27]
```diff
@@ -0,0 +1,56 @@
...
+
+#include "blocking_queue.h"
+#include <chrono>
+
+blocking_queue::blocking_queue(int buffer_size) :
+  queue_(buffer_size)
+{ }
+
+size_t blocking_queue::push(const char *arr, size_t size) {
+  std::unique_lock<std::mutex> lock(mutex_);
```
Comment (anuraaga) [inline gr-starcoder/lib/blocking_queue.cc]: From my reading of this, I don't think we need this lock since the underlying queue is thread safe

http://en.cppreference.com/w/cpp/thread/condition_variable/notify_one

Comment (kenichi-fukushima) [reply gr-starcoder/lib/blocking_queue.cc]: Sorry for overlooking this. The example code in the document release the lock before calling `notify_one`. It seems reasonable to me because the waiting thread should return from `wait` with the lock acquired. I'm not  very sure what happens if another thread has been holding the lock when it calls `notify`. The documentation says _The notifying thread does not need to hold the lock on the same mutex as the one held by the waiting thread(s); in fact doing so is a pessimization, since the notified thread would immediately block again, waiting for the notifying thread to release the lock_
